From 619c17e0dc3d1cbfdf3859b18c9e71ec00694f9c Mon Sep 17 00:00:00 2001
From: anonym <anonym@riseup.net>
Date: Wed, 27 Feb 2019 10:59:33 +0100
Subject: [PATCH] Add pref for whether to accept plaintext protocols during
 autoconfiguration.

Let's make it possible for security-focused distributions (and
extensions like TorBirdy) to prevent insecure configurations to ever
be displayed to users; for other users there is a warning explaining
the consequences of accepting a non-SSL configuration.

---
 .../components/accountcreation/content/readFromXML.js  | 10 ++++++++++
 comm/mailnews/mailnews.js                              |  6 ++++++
 2 files changed, 16 insertions(+)

diff --git a/comm/mail/components/accountcreation/content/readFromXML.js b/comm/mail/components/accountcreation/content/readFromXML.js
--- a/comm/mail/components/accountcreation/content/readFromXML.js
+++ b/comm/mail/components/accountcreation/content/readFromXML.js
@@ -34,6 +34,8 @@ function readFromXML(clientConfigXML) {
   }
   var allow_oauth2 =
     Services.prefs.getBoolPref("mailnews.auto_config.account_constraints.allow_oauth2");
+  var ssl_only =
+    Services.prefs.getBoolPref("mailnews.auto_config.sslOnly");
   var exception;
   if (
     typeof clientConfigXML != "object" ||
@@ -113,6 +115,10 @@ function readFromXML(clientConfigXML) {
       }
       exception = null;

+      if (ssl_only && iO.socketType == 1) {
+        continue;
+      }
+
       for (let iXauth of array_or_undef(iX.$authentication)) {
         try {
           iO.auth = sanitize.translate(iXauth, {
@@ -255,6 +261,10 @@ function readFromXML(clientConfigXML) {
       }
       exception = null;

+      if (ssl_only && oO.socketType == 1) {
+        continue;
+      }
+
       for (let oXauth of array_or_undef(oX.$authentication)) {
         try {
           oO.auth = sanitize.translate(oXauth, {
diff --git a/comm/mailnews/mailnews.js b/comm/mailnews/mailnews.js
--- a/comm/mailnews/mailnews.js
+++ b/comm/mailnews/mailnews.js
@@ -962,6 +962,12 @@ pref("mailnews.auto_config.guess.sslOnly", false);
 pref("mailnews.auto_config.guess.timeout", 10);
 // Whether we allow fetched configurations using OAuth2.
 pref("mailnews.auto_config.account_constraints.allow_oauth2", true);
+// Whether we allow fetched account configurations that employs
+// non-SSL/TLS protocols. With this option set, insecure
+// configurations are never presented to the user; with this option
+// unset, users picking an insecure configuration will get a warning
+// and have to opt-in.
+pref("mailnews.auto_config.sslOnly", false);
 // Work around bug 1454325 by disabling mimetype mungling in XmlHttpRequest
 pref("dom.xhr.standard_content_type_normalization", false);
